ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/DynamicServerListLoadBalancer.java

```diff
@@ -234,73 +205,23 @@ public void forceQuickPing() {
      * feature enabled
      */
     public void enableAndInitLearnNewServersFeature() {
-        keepServerListUpdated();
-        serverRefreshEnabled = true;
+        LOGGER.info("Using serverListUpdater {}", serverListUpdater.getClass().getSimpleName());
+        serverListUpdater.start(updateAction);
     }
 
     private String getIdentifier() {
         return this.getClientConfig().getClientName();
     }
 
-    private void keepServerListUpdated() {
-        scheduledFuture = _serverListRefreshExecutor.scheduleAtFixedRate(
-                new ServerListRefreshExecutorThread(),
-                LISTOFSERVERS_CACHE_UPDATE_DELAY, refeshIntervalMills,
-                TimeUnit.MILLISECONDS);
-    }
-
-    private static void shutdownExecutorPool() {
-        if (_serverListRefreshExecutor != null) {
-            _serverListRefreshExecutor.shutdown();
-
-            if (_shutdownThread != null) {
-                try {
-                    Runtime.getRuntime().removeShutdownHook(_shutdownThread);
-                } catch (IllegalStateException ise) { // NOPMD
-                    // this can happen if we're in the middle of a real
-                    // shutdown,
-                    // and that's 'ok'
-                }
-            }
-
-        }
-    }
-
     public void stopServerListRefreshing() {
-        serverRefreshEnabled = false;
-        if (scheduledFuture != null) {
-            scheduledFuture.cancel(true);
-        }
+        serverListUpdater.stop();
     }
-    
-    /**
-     * Class that updates the list of Servers This is based on the method used
-     * by the client * Appropriate Filters are applied before coming up with the
-     * right set of servers
-     * 
-     * @author stonse
-     * 
-     */
-    class ServerListRefreshExecutorThread implements Runnable {
 
-        public void run() {
-            if (!serverRefreshEnabled) {
-                if (scheduledFuture != null) {
-                    scheduledFuture.cancel(true);
-                }
-                return;
-            }
-            try {
-                updateListOfServers();
-
-            } catch (Throwable e) {
-                LOGGER.error(
-                        "Exception while updating List of Servers obtained from Discovery client",
-                        e);
-                // e.printStackTrace();
-            }
+    class ServerListRefreshAction implements ServerListUpdater.UpdateAction {
```

Comment:
This is only used in one place.  Just create an anonymous class where it's used.